test/extended/images/resolve.go

```diff
@@ -128,6 +132,325 @@ var _ = g.Describe("[Feature:ImageLookup][registry][Conformance] Image policy",
+				Name: "resolve-template",
+			},
+			Spec: kbatchv1beta1.CronJobSpec{
+				Schedule: "1 0 * * *",
+				JobTemplate: kbatchv1beta1.JobTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{
+						Annotations: map[string]string{
+							"alpha.image.policy.openshift.io/resolve-names": "*",
+						},
+					},
+					Spec: kbatchv1.JobSpec{
+						Template: kapiv1.PodTemplateSpec{
+							Spec: kapiv1.PodSpec{
+								TerminationGracePeriodSeconds: &one,
+								RestartPolicy:                 kapiv1.RestartPolicyNever,
+								Containers: []kapiv1.Container{
+									{Name: "resolve", Image: "busybox:latest", Command: []string{"/bin/true"}},
+								},
+							},
+						},
+					},
+				},
+			},
+		})
+		o.Expect(err).NotTo(o.HaveOccurred())
+		// FIXME: this needs actually fixing the admission first
+		o.Expect(cronjob.Spec.JobTemplate.Spec.Template.Spec.Containers[0].Image).NotTo(o.BeEmpty())
+		// o.Expect(cronjob.Spec.JobTemplate.Spec.Template.Spec.Containers[0].Image).To(o.Equal(tag.Image.DockerImageReference))
+	})
+
+	g.It("should perform lookup when the Deployment gets the resolve-names annotation later", func() {
+		err := oc.Run("import-image").Args("busybox:latest", "--confirm").Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		tag, err := oc.ImageClient().ImageV1().ImageStreamTags(oc.Namespace()).Get("busybox:latest", metav1.GetOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		deployment, err := oc.KubeClient().AppsV1().Deployments(oc.Namespace()).Create(&appsv1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "resolve",
+			},
+			Spec: appsv1.DeploymentSpec{
+				Selector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{"resolve": "true"},
+				},
+				Template: kapiv1.PodTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: map[string]string{"resolve": "true"},
+					},
+					Spec: kapiv1.PodSpec{
+						TerminationGracePeriodSeconds: &one,
+						Containers: []kapiv1.Container{
+							{Name: "resolve", Image: "busybox:latest", Command: []string{"/bin/true"}},
+						},
+					},
+				},
+			},
+		})
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(deployment.Spec.Template.Spec.Containers[0].Image).To(o.Equal("busybox:latest"))
+
+		g.By("auto replacing local references on Deployments when the annotation is added")
+		if deployment.Annotations == nil {
+			deployment.Annotations = map[string]string{}
+		}
+		deployment.Annotations["alpha.image.policy.openshift.io/resolve-names"] = "*"
+		deployment, err = oc.KubeClient().AppsV1().Deployments(deployment.Namespace).Patch(deployment.Name, types.StrategicMergePatchType, []byte(`{"metadata": {"annotations": {"alpha.image.policy.openshift.io/resolve-names": "*"}}}`))
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(deployment.ObjectMeta.Annotations["alpha.image.policy.openshift.io/resolve-names"]).To(o.Equal("*"))
+		o.Expect(deployment.Spec.Template.Spec.Containers[0].Image).To(o.Equal(tag.Image.DockerImageReference))
```

Comment:
@soltysh this will fail and prove https://github.com/openshift/origin/blob/8fd3eb390bdd2e1c6057fda7a3ae82c2d2fb2c6d/vendor/github.com/openshift/apiserver-library-go/pkg/admission/imagepolicy/imagereferencemutators/pods.go#L207 is broken because it doesn't respect changing the annotation which is the real reason the deployment in https://bugzilla.redhat.com/show_bug.cgi?id=1801095 never gets the image replaced and hotloops - because it sets the annotation only after creation.

I'll fix the admission tomorrow and vendor it here.

Reply:
https://github.com/openshift/apiserver-library-go/pull/24